### ISSUE TITLE
Incorrect union on small example

### ISSUE THREAD

**Issue by agersant** [21re/rust-geo-booleanop#12]
Hi and thanks for maintaining this very useful library.

I ran into some incorrect results while computing the union of a large number of polygons, and I think I managed to isolate a simple example out of my larger data set.

This is the two polygons (blue and green) being merged. Note the shared horizontal edge in the middle.
![image](https://user-images.githubusercontent.com/817256/75326826-2d4e4480-5830-11ea-9570-44a8b9e3ce78.png)

Here is the resulting union:

![image](https://user-images.githubusercontent.com/817256/75326885-49ea7c80-5830-11ea-851c-52b3ac8baa66.png)

Repro code:

```
#[macro_use]
extern crate geo_types;

use geo_booleanop::boolean::BooleanOp;

fn main() {
    let big = polygon![
        (x: 416., y: 256.),
        (x: 432., y: 240.),
        (x: 432., y: 224.),
        (x: 448., y: 280.),
        (x: 480., y: 208.),
        (x: 480., y: 256.),
    ];

    let small = polygon![
        (x: 400., y: 272.),
        (x: 416., y: 256.),
        (x: 480., y: 256.),
        (x: 480., y: 272.),
    ];

    let union = small.union(&big);
    for p in union.into_iter() {
        dbg!(p);
    }
}

/*
Output:

[src\main.rs:27] p = Polygon {
    exterior: LineString(
        [
            Coordinate {
                x: 400.0,
                y: 272.0,
            },
            Coordinate {
                x: 416.0,
                y: 256.0,
            },
            Coordinate {
                x: 432.0,
                y: 240.0,
            },
            Coordinate {
                x: 432.0,
                y: 224.0,
            },
            Coordinate {
                x: 441.14285714285717,
                y: 256.0,
            },
            Coordinate {
                x: 458.66666666666663,
                y: 256.0,
            },
            Coordinate {
                x: 480.0,
                y: 208.0,
            },
            Coordinate {
                x: 480.0,
                y: 256.0,
            },
            Coordinate {
                x: 480.0,
                y: 272.0,
            },
            Coordinate {
                x: 451.55555555555554,
                y: 272.0,
            },
            Coordinate {
                x: 448.0,
                y: 280.0,
            },
            Coordinate {
                x: 445.7142857142857,
                y: 272.0,
            },
            Coordinate {
                x: 400.0,
                y: 272.0,
            },
        ],
    ),
    interiors: [],
}
*/
```

**Comment by bluenote10** [comment]
Hi @agersant , I'm currently working on making the algorithm more robust, so further test cases are more than welcome.

I was running this example through the test case plotting:

![image](https://user-images.githubusercontent.com/3620703/75377184-94700580-58d1-11ea-989c-de1fb3a31798.png)

So it looks like there is already a coordinate hiccup in the input. Comparing to the numbers in the screenshot I noticed that there is one y value of 280 which probably should be 208. Changing it gives:

![image](https://user-images.githubusercontent.com/3620703/75377396-09dbd600-58d2-11ea-9a36-1b20356e4150.png)

If you have trouble isolating your error, feel free to post the coordinates of the full problem, preferably just as list of coordinates or GeoJSON.

I'm currently working on further fixes e.g. for https://github.com/w8r/martinez/issues/110 and https://github.com/w8r/martinez/issues/114, maybe that also helps in your case.

**Comment by agersant** [comment]
Sorry about that, I'm a potato and I shouldn't write bug reports late at night.

It is difficult to post my entire dataset as the issue arise within a iterative process which merges and simplifies a large number of polygon into gradually bigger chunks. I did take a closer look and I am confident that the union() I see are correct - I believe simplify() is where problems are. I will open an issue over there (after carefully proof reading it).

**Comment by agersant** [comment]
Sorry for all the flip flopping on this. I think I finally cornered it.

Here is a zip file with two serialized MultiPolygons (`A.json` and `B.json`). `A` is a fairly complex and one of its polygons is the blue polygon from my original example. `B` is the green polygon from the original example.

The zip file also contains `union.json` which is the union of A and B. Notice that (400, 272) is absent from the result even though it was in B.

[polys.zip](https://github.com/21re/rust-geo-booleanop/files/4259599/polys.zip)

Some pictures:

A:
![A](https://user-images.githubusercontent.com/817256/75415015-2e8a7a80-58df-11ea-84bb-a8d434d9814c.png)

B:
![B](https://user-images.githubusercontent.com/817256/75415012-2df1e400-58df-11ea-9c05-fc912eab254c.png)

Union:
![union](https://user-images.githubusercontent.com/817256/75415013-2e8a7a80-58df-11ea-9504-aca29a7f1431.png)

**Comment by bluenote10** [comment]
No problem at all, and thanks again for investigating. I can reproduce the issue now:

![image](https://user-images.githubusercontent.com/3620703/75475156-95b83580-5998-11ea-8ce2-4ecf156690ed.png)

Also, the problem seems to be independent of the issues I'm currently fixing, so it is quite a valuable test case. I'll probably do some deeper debugging on the weekend.

**Comment by agersant** [comment]
I think you are correct. I do run a simplification step on the intermediate results for performance benefits, and I only run into all these issues when using `simplify` as opposed to `simplifyvw`. With `simplifyvw`, the weird 400 -> 416 segment gets removed and the 400, 272 vertex isn't lost during subsequent unions.

I also isolated the union operation which creates this strange segment in the first place (in earlier iterations), this might be a useful test case too. Here is another zip file with `A.json` (a rectangle), `B.json` (an adjacent triangle) and `union.json` (the resulting union).
[doodad.zip](https://github.com/21re/rust-geo-booleanop/files/4265081/doodad.zip)

Picture of the resulting union w/ self-overlapping edge:

![8 5 - 6, 0 - 1, 0](https://user-images.githubusercontent.com/817256/75506507-11af7f00-5993-11ea-8d30-4db091063577.png)

**Comment by bluenote10** [comment]
Okay this is "almost" fixed by #13: The original test case works now, and also the doodad case (I had almost the same on my branch already ;)).

The thing that doesn't work is to swap the roles of the subject and clipping polygon in your main example. Will have to look into that in the next bug fixing round.

**Comment by agersant** [comment]
Fantastic news, thank you so much!!
